File: stacks-queues/155_min_stack/155_min_stack.java
/*
LeetCode 155. Min Stack
Link: https://leetcode.com/problems/min-stack/
Difficulty: Medium
Primary: Stack
Tags: [design, stack, data-structure, optimization]
Solved: 2025-11-12
*/

import java.util.*;

// --- Approach 1: One Stack ---
    // Time: O(1), Space: O(n)
class MinStack_OneStack {

    private Stack<int[]> stack;

    public MinStack_OneStack() {
        stack = new Stack<>();
    }
    
    public void push(int val) {
        int min_val;
        if (!stack.isEmpty()) {
            min_val = Math.min(stack.peek()[1], val);
        } else {
            min_val = val;
        }
        stack.push(new int[]{val, min_val});
                
    }
    
    public void pop() {
        stack.pop();
    }
    
    public int top() {
        return stack.peek()[0];
    }
    
    public int getMin() {
        return stack.peek()[1];
    }


    // --- Example Test ---
    public static void main(String[] args) {
        MinStack_OneStack sol = new MinStack_OneStack();
        
        sol.push(-2);
        sol.push(0);
        sol.push(-3);
        System.out.println(sol.getMin()); // return -3
        sol.pop();
        System.out.println(sol.top());    // return 0
        System.out.println(sol.getMin()); // return -2

    }
}

File: stacks-queues/155_min_stack/155_min_stack.py
"""
LeetCode 155. Min Stack
Link: https://leetcode.com/problems/min-stack/
Difficulty: Medium
Primary: Stack
Tags: [design, stack, data-structure, optimization]
Solved: 2025-11-12
"""

from typing import List

# --- Approach 1: One Stack ---
# Time: O(1), Space: O(n)
# each node in the stack having a minimum value 
class MinStack_OneStack:

    def __init__(self):
        self.stack = []

    def push(self, val: int) -> None:
        min_val = self.stack[-1][1] if self.stack else float('inf')
        if val < min_val: min_val = val
        self.stack.append((val, min_val))

    def pop(self) -> None:
        self.stack.pop()

    def top(self) -> int:
        return self.stack[-1][0]

    def getMin(self) -> int:
        return self.stack[-1][1]
    
# --- Approach 2: Two Stack ---
# Time: O(1), Space: O(n)
# each node in the stack having a minimum value 
class MinStack_TwoStack:

    def __init__(self):
        self.stack = []
        self.min_stack = []

    def push(self, val: int) -> None:
        min_val = self.min_stack[-1] if self.min_stack else float('inf')
        min_val = min(min_val, val)
        self.stack.append(val)
        self.min_stack.append(min_val)

    def pop(self) -> None:
        self.stack.pop()
        self.min_stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min_stack[-1]


# --- Example Test ---
if __name__ == "__main__":
    minStack = MinStack_OneStack()

    minStack.push(-2)
    minStack.push(0)
    minStack.push(-3)
    assert minStack.getMin() == -3  # return -3
    minStack.pop()
    assert minStack.top() == 0      # return 0
    assert minStack.getMin() == -2  # return -2

    print("All tests passed.")